// TODO: Copyright

syntax = "proto3";

option go_package = "nnf.cray.hpe.com/datamovement/api";
option java_multiple_files = true;
option java_package = "com.hpe.cray.nnf.datamovement";
option java_outer_classname = "DataMovementProto";

package datamovement;

// DataMover service definition describes the API for perform data movement
// for NNF storage. 
service DataMover {
    // Create sends a new data movement request identified by source and destination fields. It returns
    // a response containing a unique identifier which can be used to query the status of the request.
    rpc Create (DataMovementCreateRequest) returns (DataMovementCreateResponse) {}

    // Status requests the status of a previously submitted data movement request. It accepts a unique
    // identifier that identifies the request and returns a status message.
    rpc Status (DataMovementStatusRequest) returns (DataMovementStatusResponse) {}

    // Delete will attempt to delete a completed data movement request. It accepts a unique identifer
    // that identifies the request and returns the status of the delete operation.
    rpc Delete (DataMovementDeleteRequest) returns (DataMovementDeleteResponse) {}
}

// The data movemement create request message containing the source and destination files or directories. The
// NNF Data Mover will perform a copy from source to the destination. Specify dryrun to instantiate
// a new data movement request where the copy is simulated and not executed
message DataMovementCreateRequest {
    string workflow = 3;
    string namespace = 4;
    string source = 5;
    string destination = 6;
    bool dryrun = 7;
}

// The data movement create response message contains a unique identifier amont all data movement requests for
// the lifetime of the active job. The UID can be used to query for status of the request using the
// data movememnt status request message.
message DataMovementCreateResponse {
    string uid = 1;

    enum Status {
        CREATED = 0;
        FAILED = 1;
    }
    Status status = 2;
    string message = 3;
}

// The data movement status request message permits users to query the status of a previously issued
// data movement request by specifying the request's unique identifier.
message DataMovementStatusRequest {
    string uid = 1;
    int64 maxWaitTime = 2;
}

// The data movement status response message defines the current status of a data movement request. The
// state field describes the current state of the request. The message field contains ancillary information
// describing the data movement request. 
message DataMovementStatusResponse {
    enum State {
        PENDING = 0;
        STARTING = 1;
        RUNNING = 2;
        COMPLETED = 3;
        UNKNOWN_STATE = 4;
    }
    State state = 1;

    enum Status {
        INVALID = 0;
        NOT_FOUND = 1;
        SUCCESS = 2;
        FAILED = 3;
        UNKNOWN_STATUS = 4;
    }
    Status status = 2;
    string message = 3;
}

// The data movement delete request permits users to delete a completed data movement request (any request with a 
// status of COMPLETED) by specifying the request's unique identifier.
message DataMovementDeleteRequest {
    string uid = 1;
}

// The data movement delete response returns the status of the delete operation.
message DataMovementDeleteResponse {
    enum Status {
        INVALID = 0;
        NOT_FOUND = 1;
        DELETED = 2;
        ACTIVE = 3;
        UNKNOWN = 4;
    }
    Status status = 1;
}